libsolidity/analysis/OverflowChecker.cpp

```diff
@@ -0,0 +1,220 @@
+#include <libsolidity/interface/ErrorReporter.h>
+#include <iostream>
+#include <typeinfo>
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+using namespace boost::multiprecision;
+
+bool OverflowChecker::checkOverflow(ASTNode const& _contract)
+{
+	try
+	{
+		_contract.accept(*this);
+	}
+	catch (FatalError const&)
+	{
+		// We got a fatal error which required to stop further checking, but we can
+		// continue normally from here.
+		if (m_errorReporter.errors().empty())
+			throw; // Something is weird here, rather throw again.
+	}
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
+}
+
+bool OverflowChecker::visit(Assignment const& _assignment)
+{
+	_assignment.rightHandSide().accept(*this);
+	auto const identifier = dynamic_cast<Identifier const*>(&_assignment.leftHandSide());
+	if (identifier)
+	{
+		shared_ptr<OverflowBounds>bounds = m_overflowInfo[&_assignment.rightHandSide()];
+		auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+		m_varOverflowInfo[var] = bounds;
+		if (!bounds->fitsType(*(var->annotation().type)))
+			m_errorReporter.warning(_assignment.location(), "Variable assignment may cause overflow.  " + bounds->toString());
+		else
+			m_errorReporter.warning(_assignment.location(), "Variable assignment ok.  Range: "+bounds->toString()+" Storage: "+var->annotation().type->toString(false));
+	}
+
+	return false;
+}
+
+bool OverflowChecker::lockConstantExpression(Expression const& _expression)
+{
+	auto const* rationalNumberType = dynamic_cast<RationalNumberType const*>(_expression.annotation().type.get());
+	if (rationalNumberType)
+	{
+		shared_ptr<IntegerOverflowBounds> bounds = make_shared<IntegerOverflowBounds>(*rationalNumberType);
+		m_overflowInfo[&_expression] = bounds;
+		return true;
+	}
+	return false;
+}
+
+bool OverflowChecker::visit(Literal const& _literal)
+{
+	lockConstantExpression(_literal);
+	return true;
+}
+
+bool OverflowChecker::visit(BinaryOperation const& _operation)
+{
+	if (lockConstantExpression(_operation))
+		return false; // Do not descend
+	else
+	{
+		// Need to do more work...
+		auto const* integerType = dynamic_cast<IntegerType const*>(_operation.annotation().commonType.get());
+		if (integerType)
+		{
+			//First, descend and resolve both sides of the operation.
+			_operation.leftExpression().accept(*this);
+			_operation.rightExpression().accept(*this);
+
+			//If the operator is a comparison, then generate any possible constraints.
+			//The most naive form of this will only work if the LHS is an identifier.
+			if (Token::isCompareOp(_operation.getOperator()))
+			{
+				auto const identifier = dynamic_cast<Identifier const*>(&_operation.leftExpression());
+				if (identifier)
+				{
+					auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+					if (var)
+					{
+						IntegerOverflowBounds constraint(*_operation.leftExpression().annotation().type);
```

Comment:
I agree that constraints require a redesign, and I actually think that we'll ultimately need some sort of worker class that can generate them.  I'm just not 100% sure how I want that to work right now.

Really, ranges need a redesign, too, because they can't support being disjoint, but I'm not sure how I want that to work either (possibly with range objects being able to be composed of multiple ranges, too).